sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/communication/Channel.groovy

```diff
@@ -0,0 +1,45 @@
+package life.qbic.portal.sampletracking.communication
+/**
+ * A channel used to publish to that informs subscribers about publications
+ *
+ * @param <T> the type of data that can be moved over the channel
+ * @since 1.0.0
+ */
+class Channel<T> {
+
+    protected final Collection<Subscriber<T>> subscribers
+
+    Channel() {
+        this.subscribers = new LinkedList<>()
+    }
+
+    /**
+     * Subscribes to service update events.
+     *
+     * @param subscriber The subscriber to register for update events
+     */
+    void subscribe(Subscriber<T> subscriber) {
+        subscribers.add(subscriber)
+    }
+
+    /**
+     * Unsubscribe from the service events.
+     *
+     * @param subscription The subscription to remove
+     */
+    void unsubscribe(Subscriber<T> subscriber) {
+        subscribers.remove(subscriber)
+    }
+
+    /**
+     * Sends a message to all subscribed subscribers
+     *
+     * @param message the message to be sent
+     * @since 1.0.0
+     */
+    void publish(T message) {
+        for (Subscriber<T> subscriber : subscribers) {
+            subscriber.receive(message)
+        }
+    }
```

Comment:
I think it is a good idea to introduce interfaces here. The subscribers should not be able to publish messages. The publisher can subscribe though, nothing speaks against that.

Or is it by design, that every client can publish and receive?

Reply:
I did not want to restrict this as long as we do not require it. The publish method is not reachable when using the resource services and I am not sure where the channel might be used apart from that. I think it is fine that everyone (being a subscriber or not) is able to publish.